### PR TITLE
feat: less than component

### PR DESCRIPTION
**Pull request by esciiee**
Less Than Component:

- Compares two u64 elements
- u64 is represented as 8, 8bit M31 limbs in little endian order

## Summary by CodeRabbit

- **New Features**
  - Optimized performance with updated release configurations and integrated benchmarking.
  - Introduced advanced cryptographic modules for secure byte and numerical operations.
  - Enhanced order management with improved tracking, verification, and execution processes.
  - Upgraded hashing functionality using state-of-the-art algorithms and constants.
  - Strengthened error handling with clearer, more informative messaging.

### CODE REVIEW

File: src/components/less_than/constraints.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        eval.add_to_relation(RelationEntry::new(
+            &self.less_than_elements,
+            -E::EF::from(op.is_real.clone()),
+            &values,
+        ));
+
+        eval.finalize_logup_in_pairs();
+        eval
+    }
+}
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs]: The `evaluate` method effectively enforces all constraints for “less than” semantics. However, it is quite long and includes multiple logical steps (e.g. handling booleans, computing sums, ensuring a < b, etc.). Splitting it into smaller helper functions (e.g., “validate_flags”, “compare_bytes”, “enforce_results”) might improve readability.

File: src/executor/record.rs [hunk 2]
```diff
@@ -125,8 +135,46 @@ impl ExecutionTrace<BaseField> {
...
+                    b_cmp = *b_byte;
+                    c = BaseField::one();
+                    break;
+                }
+                std::cmp::Ordering::Greater => {
+                    *flag = BaseField::one();
+                    a_cmp = *a_byte;
+                    b_cmp = *b_byte;
+                    break;
+                }
```
Comment (rahulghangas) [inline src/executor/record.rs:167]: Currently, the inline comment states that the flags array is set to 1 for “the most significant byte where `a[i] < b[i]`", but here we're setting it for both cases

File: src/components/less_than/mod.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+impl TraceSize for LessThanColumn {
+    // last field's index + offset
+    const MAIN_COLS: usize = Self::IS_REAL + 1;
+    // 1 Interaction Column for LessThanU8 check
+    // 1 Interaction Column for yielding the result
+    const INTERACTION_COLS: usize = 2 * SECURE_EXTENSION_DEGREE;
+}
+
+relation!(LessThanElements, 18);
```
Comment (rahulghangas) [inline src/components/less_than/mod.rs]: How was this determined? Please document this

File: src/components/less_than/mod.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        assert_constraints(
+            &trace_polys,
+            CanonicCoset::new(log_size),
+            |eval| {
+                component.evaluate(eval);
+            },
+            interaction_claim.claimed_sum,
+        )
+    }
+}
```
Comment (rahulghangas) [inline src/components/less_than/mod.rs:154]: The test currently checks random values but does not guarantee coverage of the scenario where both operands are the same. This edge case sets no flags, and `c` should remain zero.

File: src/components/less_than/constraints.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+///   8. The Result of the Operation op.c must be 1 if a_comparision_byte is less than b_comparison_byte
+///      which is checked by looking up the result of the comparison bytes from the less_than_u8_elements
+///      from the PreProcessedBytes Table.
+///        - Multiplicity of the relation is the is_real flag.
+///        - Values are a_comparision_byte, b_comparison_byte, and c.
+///   9. Yield the Results by adding the values to the less_than_elements.   
+///        - Multiplicity of the relation is the negation of is_real flag.
+///        - Values are a, b, and c.
+///   10. Finalize the evaluation by calling `eval.finalize_logup_in_pairs()`.
+impl FrameworkEval for LessThanEval {
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs:44]: repeated `.clone()` calls could be reduced by carefully reusing references to avoid small overheads. While likely negligible here, it may matter in larger traces

File: src/components/less_than/constraints.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        // is_inequality_visited must be 0
+        // a_comparison_byte must be 0
+        // b_comparison_byte must be 0
+        // c must be 0
+        eval.add_constraint((E::F::one() - sum_flags.clone()) * is_inequality_visited.clone());
+        eval.add_constraint((E::F::one() - sum_flags.clone()) * a_comparison_byte.clone());
+        eval.add_constraint((E::F::one() - sum_flags.clone()) * b_comparison_byte.clone());
+        eval.add_constraint((E::F::one() - sum_flags.clone()) * op.c.clone());
+
+        // c must be 1 if a_comparision_byte is less than b_comparison_byte
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs]:
```suggestion
        // c must be 1 if a_comparison_byte is less than b_comparison_byte
```

File: src/components/less_than/constraints.rs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        // c must be 0 if a_comparision_byte is greater than b_comparison_byte
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs]:
```suggestion
        // c must be 0 if a_comparison_byte is greater than b_comparison_byte
```

File: src/components/less_than/trace.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    let mut col_gen = logup_gen.new_col();
+    for vec_row in 0..(1 << (log_size - LOG_N_LANES)) {
+        let a_comparison_byte = a_comparison_byte_col[vec_row];
+        let b_comparison_byte = b_comparison_byte_col[vec_row];
+        let a: [PackedBaseField; N_U64_LIMBS] = array::from_fn(|i| a_col[i][vec_row]);
+        let b: [PackedBaseField; N_U64_LIMBS] = array::from_fn(|i| b_col[i][vec_row]);
+        let c: PackedBaseField = c_col[vec_row];
+
+        // LessThanU8Elements Values, LookUp ensures that result is a_comparision_byte < b_comparison_byte is c.
```
Comment (rahulghangas) [inline src/components/less_than/trace.rs]:
```suggestion
        // LessThanU8Elements Values, LookUp ensures that result is a_comparison_byte < b_comparison_byte is c.
```

File: src/executor/record.rs [hunk 9]
```diff
@@ -31,50 +33,58 @@ pub type PoseidonEvents = Vec<[BaseField; 158]>;
...
-    /// Byte Events for U8 Operations
-    pub byte_operations: ByteEvents,
+    pub instructions: Vec<Instruction<F>>,
+    /// Auxillary Operations that are Looked up by the above instructions
+    /// Add Operation for Field Representations of Price, Time
+    pub add_operations: Vec<[F; 31]>,
+    /// Less Than Operation. Compares Price pairs, comprising 8 Field Elements each
+    pub less_than_operations: LessThanOperations,
+    /// Comparison Operations. Compares two Price, Time pairs
+    pub comparision_operations: Vec<[F; 53]>,
```
Comment (rahulghangas) [inline src/executor/record.rs]:
```suggestion
    pub comparison_operations: Vec<[F; 53]>,
```

File: src/components/less_than/constraints.rs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+///   1. Retrieve LessThanOp for the row
+///   2. IsReal must be a boolean. it is true if the operation is from non-padded row.
+///   3. Each flag must be a boolean. The sum of flags must be either 0 or 1.
+///      Flag represents an array of booleans, has one elment set to true if a_byte is not equal to b_byte.
+///   4. The sum of flags must be either 0 or 1. 0 if a is equal to b, 1 if a is not equal to b.
+///   5. Loop through each byte of a and b, and add constraints:
+///        if inequality is not visited yet then a_byte must be equal to b_byte and record the first byte where a is not equal to b.
+///   6. Add constraints for a_comparison_byte and b_comparison_byte.
+///   7. The Result of the Operation op.c must be a boolean.
+///   8. The Result of the Operation op.c must be 1 if a_comparision_byte is less than b_comparison_byte
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs]:
```suggestion
///   8. The Result of the Operation op.c must be 1 if a_comparison_byte is less than b_comparison_byte
```

File: src/components/less_than/constraints.rs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+///      which is checked by looking up the result of the comparison bytes from the less_than_u8_elements
+///      from the PreProcessedBytes Table.
+///        - Multiplicity of the relation is the is_real flag.
+///        - Values are a_comparision_byte, b_comparison_byte, and c.
```
Comment (rahulghangas) [inline src/components/less_than/constraints.rs]:
```suggestion
///        - Values are a_comparison_byte, b_comparison_byte, and c.
```

File: src/executor/record.rs [hunk 12]
```diff
@@ -125,8 +135,46 @@ impl ExecutionTrace<BaseField> {
     }
 
     /// Adds a Less Than Event by recording the corresponding Trace Row
-    pub fn add_less_than_event(&mut self, event: [BaseField; 26]) {
-        self.less_than_operations.push(event);
+    pub fn add_less_than_event(
```
Comment (rahulghangas) [inline src/executor/record.rs:138]: The code sets `*flag = 1` for both `a[i] < b[i]` and `a[i] > b[i]` cases, with `c = 1` only when `a[i] < b[i]`. This design effectively encodes “strictly less” with a single bit in c, while flags just marks the position of the first difference. Consider updating docstrings to reflect that this function records both “less than” and “greater than” outcomes but only sets c=1 for the less-than case

File: src/executor/record.rs [hunk 13]
```diff
@@ -125,8 +135,46 @@ impl ExecutionTrace<BaseField> {
...
+        &mut self,
+        a: [BaseField; N_U64_LIMBS],
+        b: [BaseField; N_U64_LIMBS],
+    ) {
+        let mut row = [BaseField::zero(); LessThanColumn::IS_REAL + 1];
```
Comment (rahulghangas) [inline src/executor/record.rs:143]: ### Consider short-circuit behavior for large arrays.
The current approach always iterates from the most significant limb downward, stopping at the first difference. This is fine for typical usage. However, if many events require scanning up to the last limb often, consider whether early detection or specialized instructions (e.g., an intrinsic-based comparison) might be beneficial.

**Review by rahulghangas — CHANGES_REQUESTED** [review]

File: src/components/less_than/mod.rs (hunk 4 above)
Comment (esciiee) [reply src/components/less_than/mod.rs:154]: will add a comprehensive tests of these in the generic Less Than component

File: src/components/less_than/constraints.rs (hunk 5 above)
Comment (esciiee) [reply src/components/less_than/constraints.rs:44]: EvalAtRow::F does'nt implement Copy trait, reduced the clones wherever fit

File: src/executor/record.rs (hunk 12 above)
Comment (esciiee) [reply src/executor/record.rs:138]: fixed the docs

File: src/components/less_than/constraints.rs (hunk 1 above)
Comment (esciiee) [reply src/components/less_than/constraints.rs]: I will follow up on this comment in the Generic LessThan component.

File: src/components/less_than/mod.rs (hunk 4 above)
Comment (esciiee) [reply src/components/less_than/mod.rs:154]: relayed this to [pull/12](https://github.com/aerius-labs/vex-prover/pull/12)

File: src/executor/record.rs (hunk 13 above)
Comment (rahulghangas) [reply src/executor/record.rs:143]: Not needed

File: src/executor/record.rs (hunk 2 above)
Comment (rahulghangas) [reply src/executor/record.rs:167]: @esciiee can you address this?

**Review by rahulghangas — APPROVED** [review]
LGTM
